mcuboot: Add support for swap using offset algorithm

nordicjm opened the pull request:
This new mode offers a faster update time and erase a whole round of sector erases/writes from the primary slot

~~Split MCUboot PR update due to commit being in https://github.com/zephyrproject-rtos/zephyr/pull/83294~~

nordicjm committed:
kconfig: Add new Kconfig for MCUboot swap using move

Adds a new Kconfig for selecting swap using move in MCUboot which
aligns with the proper MCUboot symbol and allows for adding
additional mode Kconfigs

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
sysbuild: Update MCUboot swap using move Kconfig name

Deprecates the old Kconfig and adds a new one which lines up with
the name of the Kconfig used by MCUboot

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
dfu: Add support for new MCUboot swap using offset mode

Allows using this newly introduced MCUboot algorithm

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
sysbuild: mcuboot: Add support for swap using offset mode

Adds a new Kconfig to allow selecting this newly introduced MCUboot
swapping algorithm

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: migration/release: 4.1: Add notes on MCUboot operating modes

Adds details on how to select this new operating mode when using
and when not using sysbuild, also includes a note about a Kconfig
being deprecated and the name of the replacement symbol

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
scripts: ci: check_compliance: Add Kconfig to allow list

Adds a Kconfig used for configuring MCUboot to the allow list

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
west.yml: MCUboot synchronization from upstream

Update Zephyr fork of MCUboot to revision:
  9a02794a14fa2e539ffe039a986360590d9e025f

Brings following Zephyr relevant fixes:
  - 9a02794a boot: bootutil: swap_offset: Skip erasing/copying
    empty sectors
  - 96b70086 zephyr: Fix image encryption configuration for
    mbedTLS.
  - e5b89529 boot: boot_serial: Add support for swap using offset
    mode
  - 18fee817 zephyr: Add support for swap using offset mode
  - 4af73169 boot: bootutil: Add swap using offset algorithm
  - 9668469f zephyr: Add Kconfig option to select PureEdDS
  - 96f01aad zephyr: Add Kconfig and configuration for SHA on
    storage
  - 30e6adfe zephyr: Allow SHA512 with TinyCrypt
  - de9bc3d0 bootutil: PureEdDSA using ED25519
  - b43b8c80 bootutil: Enable hash calculation directly on storage
  - f72158ff bootutil: Allow SHA512 with TinyCrypt
  - 4b8a37bb zephyr: Add imgtool/cbor to pip packages installed by
    west
  - 3f672134 zephyr: Fix usage of CONFIG_MBEDTLS_BUILTIN and ASN1
  - c6f5db29 bootutil: Add missing MBEDTLS_ASN1_PARSE_C

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
tests: boot: with_mcumgr: Add swap using offset test

Adds a variation of the test which tests swap using offset mode,
based on the swap using move test

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
@butok @erwango if you have boards that support swap using move and some spare time, could you check swap using offset? Can be enabled using sysbuild with `-DSB_CONFIG_MCUBOOT_MODE_SWAP_USING_OFFSET=y`

butok commented:
> @butok @erwango if you have boards that support swap using move and some spare time, could you check swap using offset? Can be enabled using sysbuild with `-DSB_CONFIG_MCUBOOT_MODE_SWAP_USING_OFFSET=y`


Testing on frdm_mcxw71 supporting swap using move.

`west build -p always -b frdm_mcxw71 --sysbuild tests/boot/test_mcuboot -- -DSB_CONFIG_MCUBOOT_MODE_SWAP_USING_OFFSET=y
`
```
Terminal log:
*** Booting MCUboot v2.1.0-rc1-219-g96b70086c8e3 ***
*** Using Zephyr OS build zephyr-v2.3.0-67834-gcae0d1c6d02e ***
I: Starting bootloader
I: Image index: 0, Swap type: none
I: Image index: 0, Swap type: none
W: Cannot upgrade: not a compatible amount of sectors
I: Bootloader chainload address offset: 0x10000
I: Image version: v0.0.0
I: Jumping to the first image slot
```
_W: Cannot upgrade: not a compatible amount of sectors._ 

So partition sizes optimized for the swap_move mode are not compatible with the swap_offset mode.

nordicjm commented:
> > @butok @erwango if you have boards that support swap using move and some spare time, could you check swap using offset? Can be enabled using sysbuild with `-DSB_CONFIG_MCUBOOT_MODE_SWAP_USING_OFFSET=y`
> 
> Testing on frdm_mcxw71 supporting swap using move.
> 
> `west build -p always -b frdm_mcxw71 --sysbuild tests/boot/test_mcuboot -- -DSB_CONFIG_MCUBOOT_MODE_SWAP_USING_OFFSET=y `
> 
> ```
> Terminal log:
> *** Booting MCUboot v2.1.0-rc1-219-g96b70086c8e3 ***
> *** Using Zephyr OS build zephyr-v2.3.0-67834-gcae0d1c6d02e ***
> I: Starting bootloader
> I: Image index: 0, Swap type: none
> I: Image index: 0, Swap type: none
> W: Cannot upgrade: not a compatible amount of sectors
> I: Bootloader chainload address offset: 0x10000
> I: Image version: v0.0.0
> I: Jumping to the first image slot
> ```
> 
> _W: Cannot upgrade: not a compatible amount of sectors._
> 
> So partition sizes optimized for the swap_move mode are not compatible with the swap_offset mode.

It's now inverse, slot 1 should be exactly 1 sector larger than slot 0 (the legacy configuration of both slots being the same size will work also)

butok commented:
> It's now inverse, slot 1 should be exactly 1 sector larger than slot 0

If I understand correctly, 1 offset sector in slot 1, but the status area is still present in slot 0, which takes >1 sector.
So they should have equal sizes (if status area takes only 1 sector).

nashif commented:
shouldn't the 2 first commits changing the manifest be squashed?

nordicjm commented:
> > It's now inverse, slot 1 should be exactly 1 sector larger than slot 0
> 
> If I understand correctly, 1 offset sector in slot 1, but the status area is still present in slot 0, which takes >1 sector. So they should have equal sizes (if status area takes only 1 sector).
> 
> So after making slot 0 and slot 1 equal:
> 
> ```
> *** Booting MCUboot v2.1.0-rc1-219-g96b70086c8e3 ***
> *** Using Zephyr OS build zephyr-v2.3.0-67834-gcae0d1c6d02e ***
> I: Starting bootloader
> I: Image index: 0, Swap type: none
> I: Image index: 0, Swap type: none
> I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
> I: Secondary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
> I: Boot source: none
> I: Image index: 0, Swap type: none
> I: Image index: 0, Swap type: none
> I: Image index: 0, Swap type: none
> I: Image index: 0, Swap type: none
> I: Bootloader chainload address offset: 0x10000
> I: Image version: v0.0.0
> I: Jumping to the first image slot
> *** Booting Zephyr OS build zephyr-v2.3.0-67834-gcae0d1c6d02e ***
> Launching primary slot application on frdm_mcxw71
> Secondary application ready for swap, rebooting*** Booting MCUboot v2.1.0-rc1-219-g96b70086c8e3 ***
> *** Using Zephyr OS build zephyr-v2.3.0-67834-gcae0d1c6d02e ***
> I: Starting bootloader
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
> I: Secondary image: magic=good, swap_type=0x3, copy_done=0x3, image_ok=0x1
> I: Boot source: none
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> E: Secondary header magic detected in first sector, wrong upload address?
> E: Image in the secondary slot is not valid!
> I: Bootloader chainload address offset: 0x10000
> I: Image version: v0.0.0
> I: Jumping to the first image slot
> *** Booting Zephyr OS build zephyr-v2.3.0-67834-gcae0d1c6d02e ***
> Launching primary slot application on frdm_mcxw71
> Secondary application ready for swap, rebooting*** Booting MCUboot v2.1.0-rc1-219-g96b70086c8e3 ***
> *** Using Zephyr OS build zephyr-v2.3.0-67834-gcae0d1c6d02e ***
> I: Starting bootloader
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
> I: Secondary image: magic=good, swap_type=0x3, copy_done=0x3, image_ok=0x1
> I: Boot source: none
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Bootloader chainload address offset: 0x10000
> I: Image version: v0.0.0
> I: Jumping to the first image slot
> *** Booting Zephyr OS build zephyr-v2.3.0-67834-gcae0d1c6d02e ***
> Launching primary slot application on frdm_mcxw71
> Secondary application ready for swap, rebooting*** Booting MCUboot v2.1.0-rc1-219-g96b70086c8e3 ***
> *** Using Zephyr OS build zephyr-v2.3.0-67834-gcae0d1c6d02e ***
> I: Starting bootloader
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
> I: Secondary image: magic=good, swap_type=0x3, copy_done=0x3, image_ok=0x1
> I: Boot source: none
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Bootloader chainload address offset: 0x10000
> I: Image version: v0.0.0
> I: Jumping to the first image slot
> ```
> 
> Errors:
> 
> ```
> E: Secondary header magic detected in first sector, wrong upload address?
> E: Image in the secondary slot is not valid!
> ```

You need to upload the update to the second sector of the update slot, not the first

> shouldn't the 2 first commits changing the manifest be squashed?

See comment from top post:

> Split MCUboot PR update due to commit being in https://github.com/zephyrproject-rtos/zephyr/pull/83294

There was a previous sync. which was included in that PR which I assume will be merged first

butok commented:
> You need to upload the update to the second sector of the update slot, not the first, run using `tests/boot/with_mcumgr`

OK. It's done buy the tests/boot/test_mcuboot.
So the test should be fixed to support the new update mode.

nordicjm committed:
tests: boot: test_mcuboot: Add swap using offset test

Adds a variation of the test which tests swap using offset mode,
based on the swap using move test

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
> > You need to upload the update to the second sector of the update slot, not the first, run using `tests/boot/with_mcumgr`
> 
> OK. It's done buy the tests/boot/test_mcuboot. So the test should be fixed to support the new update mode.

Fixed

butok commented:
> > > You need to upload the update to the second sector of the update slot, not the first, run using `tests/boot/with_mcumgr`
> > 
> > 
> > OK. It's done buy the tests/boot/test_mcuboot. So the test should be fixed to support the new update mode.
> 
> Fixed

Now, for SHA-1: b91c639b0421f1dfbaffc82c768ac373d27434b0
`west build -p always -b frdm_mcxw71 --sysbuild tests/boot/test_mcuboot -- -DSB_CONFIG_MCUBOOT_MODE_SWAP_USING_OFFSET=y `
getting error (Windows 11):

```
CMake Error at C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/yaml.cmake:326 (string):
  string sub-command JSON failed parsing json string: * Line 1, Column 40

    Missing ',' or '}' in object declaration
```


Full log:

```
C:\_ddm\ZEPHYR\TEST\test_swap_move\zephyr>west build -p always -b frdm_mcxw71 --sysbuild tests/boot/test_mcuboot -- -DSB_CONFIG_MCUBOOT_MODE_SWAP_USING_OFFSET=y
-- west build: making build dir C:\_ddm\ZEPHYR\TEST\test_swap_move\zephyr\build pristine
-- west build: generating a build system
Loading Zephyr module(s) (Zephyr base): sysbuild_default
-- Found Python3: C:/Users/nxa12205/AppData/Local/Programs/Python/Python310/python.exe (found suitable version "3.10.11", minimum required is "3.10") found components: Interpreter
-- Cache files will be written to: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/.cache
-- Found west (found suitable version "1.3.0", minimum required is "0.14.0")
-- Board: frdm_mcxw71, qualifiers: mcxw716c
Parsing C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/share/sysbuild/Kconfig
Loaded configuration 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/_sysbuild/empty.conf'
Merged configuration 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/tests/boot/test_mcuboot/sysbuild.conf'
Merged configuration 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/zephyr/misc/generated/extra_kconfig_options.conf'
Configuration saved to 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/zephyr/.config'
Kconfig header saved to 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/_sysbuild/autoconf.h'
foo: name: test_mcuboot, source-dir: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/tests/boot/test_mcuboot, type: MAIN
bar: "name":"test_mcuboot","source-dir":"C":"/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/tests/boot/test_mcuboot","type":"MAIN"
CMake Error at C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/yaml.cmake:326 (string):
  string sub-command JSON failed parsing json string: * Line 1, Column 40

    Missing ',' or '}' in object declaration

  .
Call Stack (most recent call first):
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/extensions.cmake:3782 (yaml_set)
  cmake/modules/sysbuild_images.cmake:22 (build_info)
  cmake/modules/sysbuild_images.cmake:39 (sysbuild_info_image)
  cmake/modules/sysbuild_default.cmake:21 (include)
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:75 (include)
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/share/sysbuild-package/cmake/SysbuildConfig.cmake:8 (include)
  template/CMakeLists.txt:10 (find_package)


foo: name: mcuboot, source-dir: C:/_ddm/ZEPHYR/TEST/test_swap_move/bootloader/mcuboot/boot/zephyr, type: BOOTLOADER
bar: "name":"mcuboot","source-dir":"C":"/_ddm/ZEPHYR/TEST/test_swap_move/bootloader/mcuboot/boot/zephyr","type":"BOOTLOADER"
CMake Error at C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/yaml.cmake:326 (string):
  string sub-command JSON failed parsing json string: * Line 1, Column 1

    Syntax error: value, object or array expected.

  .
Call Stack (most recent call first):
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/extensions.cmake:3782 (yaml_set)
  cmake/modules/sysbuild_images.cmake:22 (build_info)
  cmake/modules/sysbuild_images.cmake:39 (sysbuild_info_image)
  cmake/modules/sysbuild_default.cmake:21 (include)
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:75 (include)
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/share/sysbuild-package/cmake/SysbuildConfig.cmake:8 (include)
  template/CMakeLists.txt:10 (find_package)


foo: name: swapped_app, source-dir:
bar: "name":"swapped_app","source-dir":""
CMake Error at C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/yaml.cmake:326 (string):
  string sub-command JSON failed parsing json string: * Line 1, Column 1

    Syntax error: value, object or array expected.

  .
Call Stack (most recent call first):
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/extensions.cmake:3782 (yaml_set)
  cmake/modules/sysbuild_images.cmake:22 (build_info)
  cmake/modules/sysbuild_images.cmake:39 (sysbuild_info_image)
  cmake/modules/sysbuild_default.cmake:21 (include)
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:75 (include)
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/share/sysbuild-package/cmake/SysbuildConfig.cmake:8 (include)
  template/CMakeLists.txt:10 (find_package)


CMake Error at C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/yaml.cmake:411 (string):
  string sub-command JSON failed parsing json string: * Line 1, Column 1

    Syntax error: value, object or array expected.

  .
Call Stack (most recent call first):
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/yaml.cmake:392 (to_yaml)
  cmake/modules/sysbuild_images.cmake:25 (yaml_save)
  cmake/modules/sysbuild_images.cmake:39 (sysbuild_info_image)
  cmake/modules/sysbuild_default.cmake:21 (include)
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:75 (include)
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/share/sysbuild-package/cmake/SysbuildConfig.cmake:8 (include)
  template/CMakeLists.txt:10 (find_package)


--
   *****************************
   * Running CMake for mcuboot *
   *****************************

Loading Zephyr default modules (Zephyr base).
-- Application: C:/_ddm/ZEPHYR/TEST/test_swap_move/bootloader/mcuboot/boot/zephyr
-- CMake version: 3.28.1
-- Found Python3: C:/Users/nxa12205/AppData/Local/Programs/Python/Python310/python.exe (found suitable version "3.10.11", minimum required is "3.10") found components: Interpreter
-- Cache files will be written to: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/.cache
-- Zephyr version: 4.0.99 (C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr)
-- Found west (found suitable version "1.3.0", minimum required is "0.14.0")
-- Board: frdm_mcxw71, qualifiers: mcxw716c
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.17.0 (C:/Users/nxa12205/zephyr-sdk-0.17.0)
-- Found toolchain: zephyr 0.17.0 (C:/Users/nxa12205/zephyr-sdk-0.17.0)
-- Found Dtc: C:/Users/nxa12205/.mcuxpressotools/dtc-msys2/tools/usr/bin/dtc.exe (found suitable version "1.6.1", minimum required is "1.4.6")
-- Found BOARD.dts: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/boards/nxp/frdm_mcxw71/frdm_mcxw71.dts
-- Found devicetree overlay: C:/_ddm/ZEPHYR/TEST/test_swap_move/bootloader/mcuboot/boot/zephyr/app.overlay
-- Generated zephyr.dts: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/mcuboot/zephyr/zephyr.dts
-- Generated pickled edt: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/mcuboot/zephyr/edt.pickle
-- Generated devicetree_generated.h: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/mcuboot/zephyr/include/generated/zephyr/devicetree_generated.h
-- Including generated dts.cmake file: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/mcuboot/zephyr/dts.cmake

warning: MBEDTLS_CFG_FILE (defined at
C:/_ddm/ZEPHYR/TEST/test_swap_move/bootloader/mcuboot/boot/zephyr/Kconfig:288,
C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/modules/mbedtls/Kconfig:52, modules\mbedtls\Kconfig:52)
was assigned the value 'mcuboot-mbedtls-cfg.h' but got the value 'config-tls-generic.h'. See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_MBEDTLS_CFG_FILE and/or look up
MBEDTLS_CFG_FILE in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.

Parsing C:/_ddm/ZEPHYR/TEST/test_swap_move/bootloader/mcuboot/boot/zephyr/Kconfig
Loaded configuration 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/boards/nxp/frdm_mcxw71/frdm_mcxw71_defconfig'
Merged configuration 'C:/_ddm/ZEPHYR/TEST/test_swap_move/bootloader/mcuboot/boot/zephyr/prj.conf'
Merged configuration 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/tests/boot/test_mcuboot/sysbuild/mcuboot.conf'
Merged configuration 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/mcuboot/zephyr/.config.sysbuild'
Configuration saved to 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/mcuboot/zephyr/.config'
Kconfig header saved to 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/mcuboot/zephyr/include/generated/zephyr/autoconf.h'
-- Found GnuLd: c:/users/nxa12205/zephyr-sdk-0.17.0/arm-zephyr-eabi/arm-zephyr-eabi/bin/ld.bfd.exe (found version "2.38")
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: C:/Users/nxa12205/zephyr-sdk-0.17.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc.exe
Load components for MCXW716C:
driver_common component is included.
device_system component is included.
CMSIS_Include_core_cm component is included.
device_system component is included.
driver_gpio component is included.
driver_lpuart component is included.
driver_port component is included.
driver_flash_k4 component is included.
driver_ccm32k component is included.
MCUBoot bootloader key file: C:/_ddm/ZEPHYR/TEST/test_swap_move/bootloader/mcuboot/root-rsa-2048.pem
CMake Warning at CMakeLists.txt:342 (message):
  WARNING: Using default MCUboot signing key file, this file is for debug use
  only and is not secure!


Calculated maximum number of sectors: 52
-- Configuring done (12.5s)
-- Generating done (0.8s)
-- Build files have been written to: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/mcuboot
--
   *********************************
   * Running CMake for swapped_app *
   *********************************

Loading Zephyr default modules (Zephyr base).
-- Application: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/tests/boot/test_mcuboot/swapped_app
-- CMake version: 3.28.1
-- Found Python3: C:/Users/nxa12205/AppData/Local/Programs/Python/Python310/python.exe (found suitable version "3.10.11", minimum required is "3.10") found components: Interpreter
-- Cache files will be written to: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/.cache
-- Zephyr version: 4.0.99 (C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr)
-- Found west (found suitable version "1.3.0", minimum required is "0.14.0")
-- Board: frdm_mcxw71, qualifiers: mcxw716c
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.17.0 (C:/Users/nxa12205/zephyr-sdk-0.17.0)
-- Found toolchain: zephyr 0.17.0 (C:/Users/nxa12205/zephyr-sdk-0.17.0)
-- Found Dtc: C:/Users/nxa12205/.mcuxpressotools/dtc-msys2/tools/usr/bin/dtc.exe (found suitable version "1.6.1", minimum required is "1.4.6")
-- Found BOARD.dts: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/boards/nxp/frdm_mcxw71/frdm_mcxw71.dts
-- Generated zephyr.dts: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/swapped_app/zephyr/zephyr.dts
-- Generated pickled edt: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/swapped_app/zephyr/edt.pickle
-- Generated devicetree_generated.h: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/swapped_app/zephyr/include/generated/zephyr/devicetree_generated.h
-- Including generated dts.cmake file: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/swapped_app/zephyr/dts.cmake
Parsing C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/tests/boot/test_mcuboot/swapped_app/Kconfig
Loaded configuration 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/boards/nxp/frdm_mcxw71/frdm_mcxw71_defconfig'
Merged configuration 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/tests/boot/test_mcuboot/swapped_app/prj.conf'
Merged configuration 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/swapped_app/zephyr/misc/generated/extra_kconfig_options.conf'
Merged configuration 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/swapped_app/zephyr/.config.sysbuild'
Configuration saved to 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/swapped_app/zephyr/.config'
Kconfig header saved to 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/swapped_app/zephyr/include/generated/zephyr/autoconf.h'
-- Found GnuLd: c:/users/nxa12205/zephyr-sdk-0.17.0/arm-zephyr-eabi/arm-zephyr-eabi/bin/ld.bfd.exe (found version "2.38")
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: C:/Users/nxa12205/zephyr-sdk-0.17.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc.exe
Load components for MCXW716C:
driver_common component is included.
device_system component is included.
CMSIS_Include_core_cm component is included.
device_system component is included.
driver_gpio component is included.
driver_lpuart component is included.
driver_port component is included.
driver_ccm32k component is included.
-- Including signing script: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/mcuboot.cmake
-- Configuring done (12.5s)
-- Generating done (0.3s)
-- Build files have been written to: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/swapped_app
--
   **********************************
   * Running CMake for test_mcuboot *
   **********************************

Loading Zephyr default modules (Zephyr base).
-- Application: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/tests/boot/test_mcuboot
-- CMake version: 3.28.1
-- Found Python3: C:/Users/nxa12205/AppData/Local/Programs/Python/Python310/python.exe (found suitable version "3.10.11", minimum required is "3.10") found components: Interpreter
-- Cache files will be written to: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/.cache
-- Zephyr version: 4.0.99 (C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr)
-- Found west (found suitable version "1.3.0", minimum required is "0.14.0")
-- Board: frdm_mcxw71, qualifiers: mcxw716c
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.17.0 (C:/Users/nxa12205/zephyr-sdk-0.17.0)
-- Found toolchain: zephyr 0.17.0 (C:/Users/nxa12205/zephyr-sdk-0.17.0)
-- Found Dtc: C:/Users/nxa12205/.mcuxpressotools/dtc-msys2/tools/usr/bin/dtc.exe (found suitable version "1.6.1", minimum required is "1.4.6")
-- Found BOARD.dts: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/boards/nxp/frdm_mcxw71/frdm_mcxw71.dts
-- Generated zephyr.dts: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/test_mcuboot/zephyr/zephyr.dts
-- Generated pickled edt: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/test_mcuboot/zephyr/edt.pickle
-- Generated devicetree_generated.h: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/test_mcuboot/zephyr/include/generated/zephyr/devicetree_generated.h
-- Including generated dts.cmake file: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/test_mcuboot/zephyr/dts.cmake
Parsing C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/Kconfig
Loaded configuration 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/boards/nxp/frdm_mcxw71/frdm_mcxw71_defconfig'
Merged configuration 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/tests/boot/test_mcuboot/prj.conf'
Merged configuration 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/test_mcuboot/zephyr/.config.sysbuild'
Configuration saved to 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/test_mcuboot/zephyr/.config'
Kconfig header saved to 'C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/test_mcuboot/zephyr/include/generated/zephyr/autoconf.h'
-- Found GnuLd: c:/users/nxa12205/zephyr-sdk-0.17.0/arm-zephyr-eabi/arm-zephyr-eabi/bin/ld.bfd.exe (found version "2.38")
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: C:/Users/nxa12205/zephyr-sdk-0.17.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc.exe
Load components for MCXW716C:
driver_common component is included.
device_system component is included.
CMSIS_Include_core_cm component is included.
device_system component is included.
driver_gpio component is included.
driver_lpuart component is included.
driver_port component is included.
driver_flash_k4 component is included.
driver_ccm32k component is included.
-- Including signing script: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/mcuboot.cmake
-- Configuring done (12.1s)
-- Generating done (0.3s)
-- Build files have been written to: C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/build/test_mcuboot
CMake Error at C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/yaml.cmake:296 (string):
  string sub-command JSON failed parsing json string: * Line 1, Column 1

    Syntax error: value, object or array expected.

  .
Call Stack (most recent call first):
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/extensions.cmake:3782 (yaml_set)
  CMakeLists.txt:29 (build_info)


CMake Error at C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/yaml.cmake:337 (string):
  string sub-command JSON failed parsing json string: * Line 1, Column 1

    Syntax error: value, object or array expected.

  .
Call Stack (most recent call first):
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/extensions.cmake:3782 (yaml_set)
  CMakeLists.txt:29 (build_info)


CMake Error at C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/yaml.cmake:296 (string):
  string sub-command JSON failed parsing json string: * Line 1, Column 1

    Syntax error: value, object or array expected.

  .
Call Stack (most recent call first):
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/extensions.cmake:3782 (yaml_set)
  CMakeLists.txt:30 (build_info)


CMake Error at C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/yaml.cmake:337 (string):
  string sub-command JSON failed parsing json string: * Line 1, Column 1

    Syntax error: value, object or array expected.

  .
Call Stack (most recent call first):
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/extensions.cmake:3782 (yaml_set)
  CMakeLists.txt:30 (build_info)


CMake Error at C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/yaml.cmake:411 (string):
  string sub-command JSON failed parsing json string: * Line 1, Column 1

    Syntax error: value, object or array expected.

  .
Call Stack (most recent call first):
  C:/_ddm/ZEPHYR/TEST/test_swap_move/zephyr/cmake/modules/yaml.cmake:392 (to_yaml)
  CMakeLists.txt:33 (yaml_save)


-- Configuring incomplete, errors occurred!
FATAL ERROR: command exited with status 1: 'C:\Users\nxa12205\.mcuxpressotools\cmake-3.28.1-windows-x86_64\bin\cmake.EXE' -DWEST_PYTHON=C:/Users/nxa12205/AppData/Local/Programs/Python/Python310/python.exe '-BC:\_ddm\ZEPHYR\TEST\test_swap_move\zephyr\build' -GNinja -DBOARD=frdm_mcxw71 -DSB_CONFIG_MCUBOOT_MODE_SWAP_USING_OFFSET=y '-SC:\_ddm\ZEPHYR\TEST\test_swap_move\zephyr\share\sysbuild' '-DAPP_DIR:PATH=C:\_ddm\ZEPHYR\TEST\test_swap_move\zephyr\tests\boot\test_mcuboot'

C:\_ddm\ZEPHYR\TEST\test_swap_move\zephyr>

```

erwango commented:
Here's what I got:

```
$ west build -p always -b b_l4s5i_iot01a --sysbuild tests/boot/test_mcuboot -- -DSB_CONFIG_MCUBOOT_MODE_SWAP_USING_OFFSET=


*** Booting MCUboot v2.1.0-rc1-219-g96b70086c8e3 ***
*** Using Zephyr OS build v4.0.0-4126-g04fc4ba93138 ***
I: Starting bootloader
I: Image index: 0, Swap type: perm
I: Image index: 0, Swap type: perm
I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Secondary image: magic=good, swap_type=0x3, copy_done=0x3, image_ok=0x1
I: Boot source: none
I: Image index: 0, Swap type: perm
I: Image index: 0, Swap type: perm
I: Image index: 0, Swap type: perm
I: Image index: 0, Swap type: perm
E: Secondary header magic detected in first sector, wrong upload address?
E: Image in the secondary slot is not valid!
I: Bootloader chainload address offset: 0x10000
I: Image version: v0.0.0
I: Jumping to the first image slot
*** Booting Zephyr OS build v4.0.0-4126-g04fc4ba93138 ***
Launching primary slot application on b_l4s5i_iot01a
Secondary application ready for swap, rebooting
```

Using `b_l4s5i_iot01a` with `slot0_partition` reduced from 1 sector (974) vs slot1.

nordicjm commented:
> Here's what I got:
> 
> ```
> $ west build -p always -b b_l4s5i_iot01a --sysbuild tests/boot/test_mcuboot -- -DSB_CONFIG_MCUBOOT_MODE_SWAP_USING_OFFSET=
> 
> 
> *** Booting MCUboot v2.1.0-rc1-219-g96b70086c8e3 ***
> *** Using Zephyr OS build v4.0.0-4126-g04fc4ba93138 ***
> I: Starting bootloader
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
> I: Secondary image: magic=good, swap_type=0x3, copy_done=0x3, image_ok=0x1
> I: Boot source: none
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> E: Secondary header magic detected in first sector, wrong upload address?
> E: Image in the secondary slot is not valid!
> I: Bootloader chainload address offset: 0x10000
> I: Image version: v0.0.0
> I: Jumping to the first image slot
> *** Booting Zephyr OS build v4.0.0-4126-g04fc4ba93138 ***
> Launching primary slot application on b_l4s5i_iot01a
> Secondary application ready for swap, rebooting
> ```
> 
> Using `b_l4s5i_iot01a` with `slot0_partition` reduced from 1 sector (974) vs slot1.

Missing `-Dswapped_app_CONFIG_MCUBOOT_BOOTLOADER_MODE_SWAP_USING_OFFSET=y`

erwango commented:
Looks ok now if I'm correct
```
*** Booting MCUboot v2.1.0-rc1-219-g96b70086c8e3 ***
*** Using Zephyr OS build v4.0.0-4126-g04fc4ba93138 ***
I: Starting bootloader
I: Image index: 0, Swap type: none
I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Secondary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Boot source: none
I: Image index: 0, Swap type: none
I: Image index: 0, Swap type: none
I: Image index: 0, Swap type: none
I: Bootloader chainload address offset: 0x10000
I: Image version: v0.0.0
I: Jumping to the first image slot
*** Booting Zephyr OS build v4.0.0-4126-g04fc4ba93138 ***
Launching primary slot application on b_l4s5i_iot01a
Secondary application ready for swap, rebooting*** Booting MCUboot v2.1.0-rc1-219-g96b70086c8e3 ***
*** Using Zephyr OS build v4.0.0-4126-g04fc4ba93138 ***
I: Starting bootloader
I: Image index: 0, Swap type: perm
I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Secondary image: magic=good, swap_type=0x3, copy_done=0x3, image_ok=0x1
I: Boot source: none
I: Image index: 0, Swap type: perm
I: Image index: 0, Swap type: perm
I: Image index: 0, Swap type: perm
I: Starting swap using offset algorithm.
I: Image index: 0, Swap type: none
I: Image index: 0, Swap type: none
I: Image index: 0, Swap type: none
I: Bootloader chainload address offset: 0x10000
I: Image version: v0.0.0
�*** Booting Zephyr OS build v4.0.0-4126-g04fc4ba93138 ***
Swapped application booted on b_l4s5i_iot01a
```

nordicjm commented:
> Looks ok now if I'm correct
> 
> ```
> *** Booting MCUboot v2.1.0-rc1-219-g96b70086c8e3 ***
> *** Using Zephyr OS build v4.0.0-4126-g04fc4ba93138 ***
> I: Starting bootloader
> I: Image index: 0, Swap type: none
> I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
> I: Secondary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
> I: Boot source: none
> I: Image index: 0, Swap type: none
> I: Image index: 0, Swap type: none
> I: Image index: 0, Swap type: none
> I: Bootloader chainload address offset: 0x10000
> I: Image version: v0.0.0
> I: Jumping to the first image slot
> *** Booting Zephyr OS build v4.0.0-4126-g04fc4ba93138 ***
> Launching primary slot application on b_l4s5i_iot01a
> Secondary application ready for swap, rebooting*** Booting MCUboot v2.1.0-rc1-219-g96b70086c8e3 ***
> *** Using Zephyr OS build v4.0.0-4126-g04fc4ba93138 ***
> I: Starting bootloader
> I: Image index: 0, Swap type: perm
> I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
> I: Secondary image: magic=good, swap_type=0x3, copy_done=0x3, image_ok=0x1
> I: Boot source: none
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Image index: 0, Swap type: perm
> I: Starting swap using offset algorithm.
> I: Image index: 0, Swap type: none
> I: Image index: 0, Swap type: none
> I: Image index: 0, Swap type: none
> I: Bootloader chainload address offset: 0x10000
> I: Image version: v0.0.0
> �*** Booting Zephyr OS build v4.0.0-4126-g04fc4ba93138 ***
> Swapped application booted on b_l4s5i_iot01a
> ```

that's working, nice!

gchwier reviewed (dismissed):
Reviewed part with tests, looks good.
Our internal regression passed (upgrade tests on nRF boards)

nordicjm commented:
@butok that issue should be fixed now, can you try again?

butok commented:
> @butok that issue should be fixed now, can you try again?

If you are able to build it on your system, it's my local problem => ignore my issue.
`west build -p always -b frdm_mcxw71 --sysbuild tests/boot/test_mcuboot -- -DSB_CONFIG_MCUBOOT_MODE_SWAP_USING_OFFSET=y`

nordicjm commented:
> > @butok that issue should be fixed now, can you try again?
> 
> If you are able to build it on your system, it's my local problem => ignore my issue. `west build -p always -b frdm_mcxw71 --sysbuild tests/boot/test_mcuboot -- -DSB_CONFIG_MCUBOOT_MODE_SWAP_USING_OFFSET=y`

Builds here:
```
-- west build: building application
[13/24] Performing build step for 'mcuboot'
[1/285] Preparing syscall dependency handling

[4/285] Generating include/generated/zephyr/version.h
-- Zephyr version: 4.0.99 (/tmp/aa/zephyr), build: v4.0.0-4431-g8f2952484dad
[285/285] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       33006 B        64 KB     50.36%
             RAM:       21880 B       104 KB     20.55%
     RetainedMem:          0 GB         8 KB      0.00%
    rpmsg_sh_mem:          0 GB         6 KB      0.00%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /tmp/aa/zephyr/build/mcuboot/zephyr/zephyr.elf for board: frdm_mcxw71
[16/24] Performing build step for 'swapped_app'
[1/137] Preparing syscall dependency handling

[3/137] Generating include/generated/zephyr/version.h
-- Zephyr version: 4.0.99 (/tmp/aa/zephyr), build: v4.0.0-4431-g8f2952484dad
[137/137] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       15348 B       432 KB      3.47%
             RAM:        4008 B       104 KB      3.76%
     RetainedMem:          0 GB         8 KB      0.00%
    rpmsg_sh_mem:          0 GB         6 KB      0.00%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /tmp/aa/zephyr/build/swapped_app/zephyr/zephyr.elf for board: frdm_mcxw71
image.py: sign the payload
image.py: sign the payload
[19/24] Performing build step for 'test_mcuboot'
[1/153] Preparing syscall dependency handling

[3/153] Generating include/generated/zephyr/version.h
-- Zephyr version: 4.0.99 (/tmp/aa/zephyr), build: v4.0.0-4431-g8f2952484dad
[153/153] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       18414 B       424 KB      4.24%
             RAM:        4240 B       104 KB      3.98%
     RetainedMem:          0 GB         8 KB      0.00%
    rpmsg_sh_mem:          0 GB         6 KB      0.00%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /tmp/aa/zephyr/build/test_mcuboot/zephyr/zephyr.elf for board: frdm_mcxw71
image.py: sign the payload
[24/24] Completed 'test_mcuboot'
```

d3zd3z commented on subsys/mgmt/mcumgr/grp/os_mgmt/src/os_mgmt.c:
```diff
@@ -430,7 +430,10 @@ os_mgmt_mcumgr_params(struct smp_streamer *ctxt)
+#elif defined(CONFIG_MCUBOOT_BOOTLOADER_MODE_SWAP_USING_OFFSET)
+#define BOOTLOADER_MODE MCUBOOT_MODE_SWAP_USING_OFFSET
+#elif defined(CONFIG_MCUBOOT_BOOTLOADER_MODE_SWAP_USING_MOVE) || \
+defined(CONFIG_MCUBOOT_BOOTLOADER_MODE_SWAP_WITHOUT_SCRATCH)
```
This would be a lot nicer to indent the second line.  Not urgent enough, but if you have to fix it for something else.

d3zd3z reviewed (approved):
Minor issue, but doesn't warrant re-doing the patch if nothing else needs to change.
